### PR TITLE
Handle killed workers and change root span initialization in the integrations

### LINKED ISSUE #2171: [Bug]: Not seeing errors from Laravel horizon
### Bug report

I've seen some recent updated on the Laravel integration. But I still can't get my horizon process to be properly configured.

I am seeing the jobs that are being run. But I do not see the errors that are being reported there. My jobs are running successfully, but I'm using `report(new Exception(''))` to mark some stuff as failed internally. Using `report()` or `throw` in the fpm instance does seem to work properly.

I'm comparing the behaviour to that of Sentry. If both are exactly the same, we can see all of our important logs in a single tool. Because Laravel has its own exception handler, it may be a good idea to look into that.

I'm pretty sure I can write something that would sent all "Laravel" errors to Datadog, I'm just looking for a way to disable the general php error reporting. If there is anywrong with my configuration that could solve this, that would also be fine ofcourse.

Thanks in advance.

### PHP version

8.2.7

### Tracer or profiler version

0.90.0

### Installed extensions

```
[PHP Modules]
bcmath
calendar
Core
ctype
curl
date
ddtrace
dom
exif
FFI
fileinfo
filter
ftp
gettext
hash
iconv
igbinary
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcov
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
random
readline
redis
Reflection
session
shmop
SimpleXML
soap
sockets
sodium
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tokenizer
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache
ddtrace
```

### Output of `phpinfo()`

```json
{
  "date": "2023-07-19T09:40:07Z",
  "os_name": "Linux ~redacted~ SMP Tue Jun 27 21:48:55 UTC 2023 x86_64",
  "os_version": "~redacted~x86_64",
  "version": "0.90.0",
  "lang": "php",
  "lang_version": "8.2.1",
  "env": "~redacted~",
  "enabled": true,
  "service": "~redacted~-worker",
  "enabled_cli": true,
  "agent_url": "http:\/\/localhost:8126",
  "debug": false,
  "analytics_enabled": false,
  "sample_rate": 1,
  "sampling_rules": [],
  "tags": {
    "generated-by": "cloudformation",
    "monitored-by": "datadoghq.eu-commerce",
  },
  "service_mapping": {
    "pdo": "~redacted~-rds",
    "guzzle": "~redacted~-guzzle",
    "curl": "~redacted~-curl",
    "phpredis": "~redacted~-redis",
    "laravel": "~redacted~-laravel"
  },
  "distributed_tracing_enabled": true,
  "priority_sampling_enabled": true,
  "dd_version": "release_2023_07_19_111522",
  "architecture": "x86_64",
  "sapi": "fpm-fcgi",
  "datadog.trace.request_init_hook": "\/opt\/datadog\/dd-library\/0.90.0\/dd-trace-sources\/bridge\/dd_wrap_autoloader.php",
  "open_basedir_configured": false,
  "uri_fragment_regex": null,
  "uri_mapping_incoming": null,
  "uri_mapping_outgoing": null,
  "auto_flush_enabled": false,
  "generate_root_span": true,
  "http_client_split_by_domain": false,
  "measure_compile_time": true,
  "report_hostname_on_root_span": false,
  "traced_internal_functions": null,
  "auto_prepend_file_configured": false,
  "integrations_disabled": "default",
  "enabled_from_env": true,
  "opcache.file_cache": null
}
```

### Upgrading from

_No response_

### PR DESCRIPTION
**Pull request by PROFeNoM**
### Description

Fixes #2171 by:
- Force-flushing the traces before the process is [killed](https://github.com/illuminate/queue/blob/99e275f601f17b187324f10d8349476bf046b733/Worker.php#L771) by Laravel queue (`MaxAttemptsExceededException` handling, see screenshot below)
- Allowing the Laravel integration's hooked to be installed even without a root span (`report` helper function usage handling, see screenshot below); the latter is also applied to other integrations

**Screenshots**

- `report(new \Exception("Reported Exception in SendVerificationEmail"));` 
image

- Job timeout
image

### Readiness checklist
- [X] (only for Members) Changelog has been added to the release document.
- [ ] Tests added for this feature/bug.

### Reviewer checklist
- [X] Appropriate labels assigned.
- [X] Milestone is set.
- [X] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: src/Integrations/Integrations/Yii/YiiIntegration.php [hunk 1]
```diff
@@ -38,16 +40,20 @@ public function init()
...
-        }
+        $service = \ddtrace_config_app_name(YiiIntegration::NAME);
 
-        $rootSpan->meta[Tag::COMPONENT] = YiiIntegration::NAME;
-        $rootSpan->meta[Tag::SPAN_KIND] = 'server';
+        \DDTrace\hook_method(
+            'yii\di\Container',
+            '__construct',
+            function () {
+                $rootSpan = \DDTrace\root_span();
```
Comment (bwoebi) [inline src/Integrations/Integrations/Yii/YiiIntegration.php:49]: This is a hook method - check for actual root span existing.

File: src/Integrations/Integrations/Yii/YiiIntegration.php [hunk 2]
```diff
@@ -92,13 +98,18 @@ function (SpanData $span, $args) use ($service) {
...
                 $span->name = \get_class($this) . '.runAction';
                 $span->type = Type::WEB_SERVLET;
                 $span->service = $service;
                 $span->resource = YiiIntegration::extractResourceNameFromRunAction($args) ?: $span->name;
                 $span->meta[Tag::COMPONENT] = YiiIntegration::NAME;
 
+                $rootSpan = root_span();
+                if ($rootSpan === null) {
+                    return;
+                }
```
Comment (bwoebi) [inline src/Integrations/Integrations/Yii/YiiIntegration.php:111]: Unnecessary: we're in a trace method.

File: src/Integrations/Integrations/Symfony/SymfonyIntegration.php [hunk 3]
```diff
@@ -449,7 +446,8 @@ function (SpanData $span) use ($controllerName) {
                         return;
                     }
                     if (!$injectedActionInfo) {
-                        if ($integration->injectActionInfo($event, $eventName, $integration->symfonyRequestSpan)) {
+                        $rootSpan = root_span();
+                        if ($rootSpan !== null && $integration->injectActionInfo($event, $eventName, $rootSpan)) {
```
Comment (bwoebi) [inline src/Integrations/Integrations/Symfony/SymfonyIntegration.php:450]:
```suggestion
                        if ($integration->injectActionInfo($event, $eventName, root_span())) {
```
Redundant: in trace_method

File: src/Integrations/Integrations/Symfony/SymfonyIntegration.php [hunk 4]
```diff
@@ -357,23 +349,28 @@ function (SpanData $span, $args, $response) use ($integration) {
                 $span->type = Type::WEB_SERVLET;
                 $span->meta[Tag::COMPONENT] = SymfonyIntegration::NAME;
 
-                $integration->symfonyRequestSpan->meta[Tag::HTTP_METHOD] = $request->getMethod();
-                $integration->symfonyRequestSpan->meta[Tag::COMPONENT] = SymfonyIntegration::NAME;
-                $integration->symfonyRequestSpan->meta[Tag::SPAN_KIND] = 'server';
+                $rootSpan = root_span();
+                if ($rootSpan === null) {
+                    return;
+                }
```
Comment (bwoebi) [inline src/Integrations/Integrations/Symfony/SymfonyIntegration.php:355]:
```suggestion
```
Redundant: in trace_method

File: src/Integrations/Integrations/Symfony/SymfonyIntegration.php [hunk 5]
```diff
@@ -69,10 +74,6 @@ public function init()
             'boot',
             [
                 'prehook' => function (SpanData $span) {
-                    if (\DDTrace\root_span() === $span) {
-                        return false;
-                    }
```
Comment (bwoebi) [inline src/Integrations/Integrations/Symfony/SymfonyIntegration.php:74]: Why was this removed?

Comment (PROFeNoM) [reply src/Integrations/Integrations/Symfony/SymfonyIntegration.php]: You are right.

File: src/Integrations/Integrations/LaravelQueue/LaravelQueueIntegration.php [hunk 6]
```diff
@@ -46,6 +45,25 @@ public function init()
...
 
+        \DDTrace\hook_method(
+            'Illuminate\Queue\Worker',
+            'kill',
+            function () {
+                // span.c:ddtrace_close_all_open_spans() -> ensure the span is closed rather than dropped
+                ini_set('datadog.autofinish_spans', '1');
+
+                dd_trace_close_all_spans_and_flush();
+                dd_trace_internal_fn('synchronous_flush', 3);
```
Comment (bwoebi) [inline src/Integrations/Integrations/LaravelQueue/LaravelQueueIntegration.php:56]: Maybe this should be promoted to a proper API function then, if there are use cases outside of testing :-)

[the remark this answers is not in the document]
Comment (bwoebi) [reply src/Integrations/Integrations/LaravelQueue/LaravelQueueIntegration.php:56]: Yes, the _call_ is right, it should just have a proper API function for that use case.

[the remark this answers is not in the document]
Comment (PROFeNoM) [reply src/Integrations/Integrations/LaravelQueue/LaravelQueueIntegration.php:56]: Ok, I'll add it to the API then

File: src/Integrations/Integrations/Symfony/SymfonyIntegration.php (hunk 4 above)
Comment (PROFeNoM) [reply src/Integrations/Integrations/Symfony/SymfonyIntegration.php]: This check happens in a hook on `Symfony\Component\HttpKernel\HttpKernel::handle`, which is the same method as the one on line 40 of the same file. I'm not totally convinced that this check should be removed, to be 100% honest.

File: src/Integrations/Integrations/LaravelQueue/LaravelQueueIntegration.php [hunk 7]
```diff
@@ -109,12 +127,28 @@ public function init()
...
             'fire',
-            function (SpanData $span, $args, $retval, $exception) use ($integration) {
-                $integration->setSpanAttributes($span, 'laravel.queue.fire', 'process', $this, $exception);
-            }
+            [
+                'prehook' => function (SpanData $span, $args, $retval) use ($integration) {
+                    $integration->setSpanAttributes($span, 'laravel.queue.fire', 'process', $this);
+                },
+                'posthook' => function (SpanData $span, $args, $retval, $exception) use ($integration) {
+                    $integration->setSpanAttributes($span, 'laravel.queue.fire', 'process', $this, $exception);
```
Comment (bwoebi) [inline src/Integrations/Integrations/LaravelQueue/LaravelQueueIntegration.php:149]:
```suggestion
                    if ($exception) {
                        $this->setError($span, $exception);
                    }
```
Avoid duplicating setting it :-)

File: src/Integrations/Integrations/Laravel/LaravelIntegration.php [hunk 8]
```diff
@@ -255,22 +277,24 @@ function (SpanData $span) use ($rootSpan, $integration) {
...
-                $span->name = 'laravel.provider.load';
-                $span->type = Type::WEB_SERVLET;
-                $span->service = $serviceName;
-                $span->resource = 'Illuminate\Foundation\ProviderRepository::load';
-                $rootSpan->name = 'laravel.request';
-                $rootSpan->service = $serviceName;
-                $rootSpan->meta[Tag::COMPONENT] = LaravelIntegration::NAME;
-                $span->meta[Tag::COMPONENT] = LaravelIntegration::NAME;
+                $rootSpan = \DDTrace\root_span();
+                if ($rootSpan !== null) {
```
Comment (bwoebi) [inline src/Integrations/Integrations/Laravel/LaravelIntegration.php:281]: Redundant check: already in trace_method

**Comment by bwoebi** [comment]
I'm not sure about the checks for if ($rootSpan) in all these trace_method() callbacks - where the root span is guaranteed to exist...

File: src/Integrations/Integrations/Symfony/SymfonyIntegration.php [hunk 9]
```diff
@@ -281,14 +284,8 @@ function ($This, $scope) use (&$symfonyCommandsIntegrated) {
...
 
-        $rootSpan = \DDTrace\root_span();
-        if (null == $rootSpan) {
-            return Integration::NOT_LOADED;
-        }
-
         /** @var SpanData $symfonyRequestSpan */
-        $this->symfonyRequestSpan = $rootSpan;
-        $this->addTraceAnalyticsIfEnabled($rootSpan);
+        $this->symfonyRequestSpan = \DDTrace\root_span();
```
Comment (bwoebi) [inline src/Integrations/Integrations/Symfony/SymfonyIntegration.php:288]: What's the purpose of this? Every place but `Symfony\Component\HttpKernel\Event\FilterControllerEvent` and `Symfony\Component\HttpKernel\HttpKernel::handle` do access the actual \DDTrace\root_span(). What is particular about these places?

File: ext/ddtrace.c [hunk 10]
```diff
@@ -1656,6 +1656,32 @@ PHP_FUNCTION(dd_trace_peek_span_id) {
     RETURN_STR(ddtrace_span_id_as_string(ddtrace_peek_span_id()));
 }
 
+/* {{{ proto void dd_trace_close_all_spans_and_flush() */
+PHP_FUNCTION(dd_trace_close_all_spans_and_flush) {
+    UNUSED(execute_data);
```
Comment (bwoebi) [inline ext/ddtrace.c:1661]:
```suggestion
    zend_parse_parameters_none();
```
We're going to move away from "log_onceerrf" in favour of proper exceptions for invalid code. Let's do it for new code straight away.

File: ext/ddtrace.c [hunk 11]
```diff
@@ -1656,6 +1656,32 @@ PHP_FUNCTION(dd_trace_peek_span_id) {
...
+}
+
+/* {{{ proto void dd_trace_synchronous_flush(int) */
+PHP_FUNCTION(dd_trace_synchronous_flush) {
+    zend_long timeout;
+
+    if (zend_parse_parameters_ex(ddtrace_quiet_zpp(), ZEND_NUM_ARGS(), "l", &timeout) == FAILURE) {
+        ddtrace_log_onceerrf("dd_trace_synchronous_flush() expects a timeout in milliseconds");
+        RETURN_NULL();
+    }
```
Comment (bwoebi) [inline ext/ddtrace.c:1673]:
```suggestion
    if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &timeout) == FAILURE) {
        RETURN_THROWS();
    }
```

File: ext/ddtrace.stub.php [hunk 12]
```diff
@@ -818,6 +818,11 @@ function dd_trace_serialize_closed_spans(): array {}
      */
     function dd_trace_peek_span_id(): string {}
 
+    /**
+     * Force close all spans and force-send finished traces to the agent
```
Comment (bwoebi) [inline ext/ddtrace.stub.php:822]: This drops or closes the spans depending on `datadog.autofinish_spans`, right? Can we please add a note about that in the docblock here?

**Comment by bwoebi** [comment]
Generally looks good to me, nice pull request :-)

File: ext/ddtrace.c (hunk 11 above)
Comment (bwoebi) [reply ext/ddtrace.c]: RETURN_THROWS() just leaves the return value unchanged and asserts an exception happened. For < PHP 8 you probably could just polyfill it in compatibility.h with `#define RETURN_THROWS RETURN_NULL`.

**Comment by bwoebi** [comment]
More like 40%, but yes :-P

**Review by bwoebi — APPROVED** [review]

**Pull request merged**